File: cpp/ml_methods/scaler_methods/demo_mlm_scaler_01/demo_mlm_scaler_01.cpp
// ===================================================================================
// demo_mlm_scaler_01.cpp
// ===================================================================================
/**
 * @file demo_mlm_scaler_01.cpp
 *
 * @brief Demonstration of ZScoreScaler and LpNormScaler usage.
 *
 * Organization:
 * -------------
 * |-- ml_methods/
 *     |-- scaler_methods/
 *         |-- demo_mlm_scaler_01.cpp
 */
// ===================================================================================

// std includes
#include <iomanip>
#include <iostream>
#include <random>
#include <string>

// Eigen includes
#include <Eigen/Dense>

// TRex Selector includes
#include <ml_methods/scaler_methods/lp_norm_scaler.hpp>
#include <ml_methods/scaler_methods/z_score_scaler.hpp>
#include <utils/openmp/utils_openmp.hpp>
#include <utils/memmap/memory_mapped_matrix.hpp>
#include <utils/eval_metrics/utils_eval_cdiagnostics.hpp>

// ============================================================================
// Namespace aliases
// ============================================================================

namespace scaler = trex::ml_methods::scaler_methods;
namespace omp_utils = trex::utils::openmp;
namespace memmap = trex::utils::memmap;
namespace cdiagnostics = trex::utils::eval::cdiagnostics;


// ============================================================================
// Helper: aggregate column statistics
// ============================================================================

/**
 * @brief Prints compact per-column statistics, averaged over all columns.
 *
 * Reveals the effect of the center / scale switches at a glance:
 *   - centering drives the average absolute column mean toward 0,
 *   - z-score scaling drives the average column SD toward 1,
 *   - L2 normalization drives the average column L2 norm toward 1.
 */
void print_column_stats(const std::string& label, const Eigen::MatrixXd& X) {
    const Eigen::Index n = X.rows();
    const Eigen::RowVectorXd col_mean = X.colwise().mean();
    const Eigen::RowVectorXd col_sd =
        ((X.rowwise() - col_mean).colwise().squaredNorm()
             / static_cast<double>(n - 1)).cwiseSqrt();
    const Eigen::RowVectorXd col_l2 = X.colwise().norm();

    std::cout << "  " << std::left << std::setw(36) << label
              << "avg|mean| = " << std::scientific << std::setprecision(2)
              << col_mean.cwiseAbs().mean()
              << "   avg SD = " << std::fixed << std::setprecision(4) << col_sd.mean()
              << "   avg L2 = " << std::setprecision(2) << col_l2.mean() << "\n";
}


// ============================================================================
// Demo 1: center / scale switches (R scale() semantics)
// ============================================================================

void demo_scaler_comparison() {
    cdiagnostics::print_section_header("Scaler Comparison - center / scale switches");

    // Step 1. Generate data with a deliberately non-zero mean (5.0) and
    //         non-unit spread (SD 3.0), so the effect of each switch is visible.
    const Eigen::Index n = 1'000, p = 500;
    std::mt19937 rng(123);
    std::normal_distribution<double> norm_dist(5.0, 3.0);

    Eigen::MatrixXd X(n, p);
    for (Eigen::Index i = 0; i < n; ++i) {
        for (Eigen::Index j = 0; j < p; ++j) {
            X(i, j) = norm_dist(rng);
        }
    }

    std::cout << "Data: " << n << " x " << p << " drawn from N(mean=5, sd=3)\n\n";
    print_column_stats("original (untransformed)", X);
    std::cout << "\n";

    // Step 2. ZScoreScaler under all four center / scale combinations.
    //         The (center, scale) booleans follow R's scale(): centering drives
    //         avg|mean| -> 0, scaling drives avg SD -> 1. With center=false the
    //         scale statistic is the root-mean-square around 0, not the SD.
    struct ZCase {
        bool center;
        bool scale;
        const char* label;
    };

    const ZCase z_cases[] = {
        {true,  true,  "ZScore(center=true,  scale=true )"},
        {false, true,  "ZScore(center=false, scale=true )"},
        {true,  false, "ZScore(center=true,  scale=false)"},
        {false, false, "ZScore(center=false, scale=false)"},
    };
    for (const auto& c : z_cases) {
        Eigen::MatrixXd Xt = X;
        Eigen::Map<Eigen::MatrixXd> Xt_map(Xt.data(), n, p);
        scaler::ZScoreScaler zscaler(c.center, c.scale);
        zscaler.fit(Xt_map);
        zscaler.transform_inplace(Xt_map);
        print_column_stats(c.label, Xt);
    }
    std::cout << "\n";

    // Step 3. LpNormScaler: L2 vs L1, with centering on and off (scale kept on).
    //         With scale=true each column is divided by its Lp norm around the
    //         applied center, so the average column L2 norm collapses toward 1
    //         for the L2 variant.
    struct LCase {
        scaler::LpNormScaler::NormType norm;
        bool center;
        const char* label;
    };

    const LCase l_cases[] = {
        {scaler::LpNormScaler::NormType::L2, true,  "LpNorm(L2, center=true )"},
        {scaler::LpNormScaler::NormType::L2, false, "LpNorm(L2, center=false)"},
        {scaler::LpNormScaler::NormType::L1, true,  "LpNorm(L1, center=true )"},
    };
    for (const auto& c : l_cases) {
        Eigen::MatrixXd Xt = X;
        Eigen::Map<Eigen::MatrixXd> Xt_map(Xt.data(), n, p);
        scaler::LpNormScaler lscaler(c.norm, c.center, /*scale=*/true);
        lscaler.fit(Xt_map);
        lscaler.transform_inplace(Xt_map);
        print_column_stats(c.label, Xt);
    }

    std::cout << "\n\n";
}


// =============================================================================
// Demo 2: Transform and Inverse Transform Test
// =============================================================================

void demo_scaler_inverse_transform() {
    cdiagnostics::print_section_header("Scaler Inverse Transform Accuracy");

    // Step 1. Generate random data
    const Eigen::Index n = 1'000, p = 5'000;
    std::cout << "Matrix size: " << n << " x " << p << "\n\n";

    std::mt19937 rng(456);
    std::normal_distribution<double> norm_dist(0.0, 1.0);
    Eigen::MatrixXd X_orig(n, p);
    for (Eigen::Index i = 0; i < n; ++i) {
        for (Eigen::Index j = 0; j < p; ++j) {
            X_orig(i, j) = norm_dist(rng);
        }
    }

    // Step 2. Test ZScoreScaler
    {
        Eigen::MatrixXd X = X_orig;
        Eigen::Map<Eigen::MatrixXd> X_map(X.data(), n, p);
        scaler::ZScoreScaler zscaler;

        zscaler.fit(X_map);
        zscaler.transform_inplace(X_map);
        zscaler.inverse_transform_inplace(X_map);

        double max_error = (X - X_orig).cwiseAbs().maxCoeff();
        double mean_error = (X - X_orig).cwiseAbs().mean();

        std::cout << "ZScoreScaler:\n"
                  << "  Max Error: " << std::scientific << max_error << "\n"
                  << "  Mean Error: " << mean_error << "\n\n";
    }

    // Step 3. Test LpNormScaler L2
    {
        Eigen::MatrixXd X = X_orig;
        Eigen::Map<Eigen::MatrixXd> X_map(X.data(), n, p);
        scaler::LpNormScaler l2scaler(scaler::LpNormScaler::NormType::L2, true);


        l2scaler.fit(X_map);
        l2scaler.transform_inplace(X_map);
        l2scaler.inverse_transform_inplace(X_map);

        double max_error = (X - X_orig).cwiseAbs().maxCoeff();
        double mean_error = (X - X_orig).cwiseAbs().mean();

        std::cout << "LpNormScaler L2:\n"
                  << "  Max Error: " << std::scientific << max_error << "\n"
                  << "  Mean Error: " << mean_error << "\n\n";
    }
    std::cout << "\n\n";
}


// =============================================================================
// Demo 3: Serialization
// =============================================================================

void demo_scaler_serialization() {
    cdiagnostics::print_section_header("Scaler Serialization");

    // Step 1. Generate data
    const Eigen::Index n = 1'000, p = 500;

    std::mt19937 rng(999);
    std::normal_distribution<double> norm_dist(0.0, 1.0);
    Eigen::MatrixXd X(n, p);
    for (Eigen::Index i = 0; i < n; ++i) {
        for (Eigen::Index j = 0; j < p; ++j) {
            X(i, j) = norm_dist(rng);
        }
    }

    Eigen::Map<Eigen::MatrixXd> X_map(X.data(), n, p);

    // Step 2. Fit and save
#ifdef TREX_DEMO_SCALER_SER_PATH
    const std::string filename = TREX_DEMO_SCALER_SER_PATH;
#else
    const std::string filename = "scaler_test.bin";
#endif
    scaler::ZScoreScaler scaler;
    scaler.fit(X_map);
    scaler.save(filename);
    std::cout << "✓ Saved ZScoreScaler to '" << filename << "'\n";

    // Step 3. Load and verify
    scaler::ZScoreScaler loaded_scaler;
    loaded_scaler.load(filename);
    std::cout << "✓ Loaded ZScoreScaler from '" << filename << "'\n\n";

    // Verify means and scales match
    double mean_diff = (scaler.get_centers() - loaded_scaler.get_centers()).norm();
    double scale_diff = (scaler.get_scales() - loaded_scaler.get_scales()).norm();

    std::cout << "Verification:\n"
              << "  Mean difference: " << std::scientific << mean_diff << "\n"
              << "  Scale difference: " << scale_diff << "\n\n";

    if (mean_diff < 1e-10 && scale_diff < 1e-10) {
        std::cout << "✓ Serialization perfect!\n\n";
    }

    // Step 4. Clean up
    std::remove(filename.c_str());
    std::cout << "\n\n";
}


// ============================================================================
// Demo 04: Memory-Mapped Matrix
// ============================================================================

void demo_scaler_on_mmap_matrix() {
    cdiagnostics::print_section_header("Scaler on Memory-Mapped Matrix");

    // Step 1. Create memory-mapped matrix
    std::cout << "Creating memory-mapped matrix...\n";
    const Eigen::Index n = 20'000, p = 100'000;
    std::cout << "Matrix size: " << n << " x " << p << "\n";
#ifdef TREX_DEMO_MMAP_PATH
    const std::string mmap_filename = TREX_DEMO_MMAP_PATH;
#else
    const std::string mmap_filename = "mmap_matrix.bin";
#endif
    memmap::MemoryMappedMatrix<double> mmap_matrix(
        mmap_filename, n, p, memmap::AccessMode::ReadWrite);
    auto X_map = mmap_matrix.getMap();

    // Step 2. Populate matrix
    #pragma omp parallel
    {
        #pragma omp for schedule(static)
        for (Eigen::Index j = 0; j < p; ++j) {
            for (Eigen::Index i = 0; i < n; ++i) {
                X_map(i, j) = static_cast<double>(2.0 * static_cast<double>(i) +
                                                           static_cast<double>(j) / 100);
            }
        }
    }

    // Step 3. Fit and run LpNormScaler
    scaler::LpNormScaler l2scaler(scaler::LpNormScaler::NormType::L2, true);
    std::cout << "Fitting L2 LpNormScaler on memory-mapped matrix...\n";
    l2scaler.fit(X_map);
    std::cout << "✓ LpNormScaler fitted\n";
    l2scaler.transform_inplace(X_map);
    std::cout << "✓ Data transformed\n";
    l2scaler.inverse_transform_inplace(X_map);
    std::cout << "✓ Data inverse transformed\n";

    // Step 4. Clean up
    std::remove(mmap_filename.c_str());
    std::cout << "✓ Memory-mapped file removed\n";
    std::cout << "\n\n";
}



// ============================================================================
// Main
// ============================================================================

void init_omp_environment(int threads = 6) {
    omp_set_num_threads(threads);
    omp_set_schedule(omp_sched_static, 0);
}


int main() {

    std::cout << "\n";
    cdiagnostics::print_section_header("DataTransformer Demo Suite");

    // Print OpenMP info
    omp_utils::print_info();
    init_omp_environment(6);
    std::cout << "Running with " << omp_get_max_threads() << " threads\n\n";

    try {
        demo_scaler_comparison();
        demo_scaler_inverse_transform();
        demo_scaler_serialization();
        demo_scaler_on_mmap_matrix();

        std::cout << "\n";
        cdiagnostics::print_section_header("All Demos Completed Successfully");

    } catch (const std::exception& e) {
        std::cerr << "\n[ERROR] " << e.what() << "\n\n";
        return 1;
    }

    return 0;
}

File: Python/ml_methods/scaler_methods/demo_mlm_scaler_01/demo_mlm_scaler_01.py
# ==============================================================================
# demo_mlm_scaler_01.py
# ==============================================================================
#
# Demonstration of ZScoreScaler and LpNormScaler usage.
#
# Mirrors cpp/ml_methods/scaler_methods/demo_mlm_scaler_01/demo_mlm_scaler_01.cpp.
#
# Four demos which illustrate the column-scaling utilities:
#
#   Demo 1 | Scaler comparison  : ZScoreScaler under all four center / scale
#                                 combinations and LpNormScaler (L2 / L1),
#                                 summarized through aggregate column stats.
#
#   Demo 2 | Inverse transform  : fit -> transform_inplace ->
#                                 inverse_transform_inplace round-trip accuracy
#                                 for ZScoreScaler and the L2 LpNormScaler.
#
#   Demo 3 | Serialization      : save() a fitted ZScoreScaler to disk, load()
#                                 it into a fresh object, compare parameters.
#
#   Demo 4 | Memory-mapped      : L2 LpNormScaler fit / transform / inverse
#                                 transform directly on a MemoryMappedMatrix
#                                 zero-copy view.
#                                 (cpp: n = 20,000, p = 100,000, filled with an
#                                 OpenMP parallel loop; scaled to n = 2,000,
#                                 p = 10,000 here and filled with a vectorized
#                                 NumPy broadcast instead.)
#
# ==============================================================================

import os
import tempfile

import numpy as np

from trex_selector_neo.ml_methods import LpNormScaler, NormType, ZScoreScaler
from trex_selector_neo.utils import (
    AccessMode,
    MemoryMappedMatrix,
    get_max_threads,
    set_num_threads,
)


def print_section_header(title):
    print("=" * 78)
    print(f" {title}")
    print("=" * 78 + "\n")


def print_column_stats(label, X):
    """Print compact per-column statistics, averaged over all columns.

    Reveals the effect of the center / scale switches at a glance:
      - centering drives the average absolute column mean toward 0,
      - z-score scaling drives the average column SD toward 1,
      - L2 normalization drives the average column L2 norm toward 1.
    """
    col_mean = X.mean(axis=0)
    col_sd = X.std(axis=0, ddof=1)
    col_l2 = np.linalg.norm(X, axis=0)
    print(f"  {label:<36}avg|mean| = {np.abs(col_mean).mean():.2e}"
          f"   avg SD = {col_sd.mean():.4f}"
          f"   avg L2 = {col_l2.mean():.2f}")


# ==============================================================================
# Demo 1: center / scale switches (R scale() semantics)
# ==============================================================================

def demo_scaler_comparison():
    print_section_header("Scaler Comparison - center / scale switches")

    # Step 1. Generate data with a deliberately non-zero mean (5.0) and
    #         non-unit spread (SD 3.0), so the effect of each switch is visible.
    n, p = 1_000, 500
    rng = np.random.default_rng(123)
    X = np.asfortranarray(rng.normal(5.0, 3.0, (n, p)))

    print(f"Data: {n} x {p} drawn from N(mean=5, sd=3)\n")
    print_column_stats("original (untransformed)", X)
    print()

    # Step 2. ZScoreScaler under all four center / scale combinations.
    #         The (center, scale) booleans follow R's scale(): centering drives
    #         avg|mean| -> 0, scaling drives avg SD -> 1. With center=False the
    #         scale statistic is the root-mean-square around 0, not the SD.
    z_cases = [
        (True,  True,  "ZScore(center=True,  scale=True )"),
        (False, True,  "ZScore(center=False, scale=True )"),
        (True,  False, "ZScore(center=True,  scale=False)"),
        (False, False, "ZScore(center=False, scale=False)"),
    ]
    for center, scale, label in z_cases:
        Xt = X.copy(order="F")
        zscaler = ZScoreScaler(center=center, scale=scale)
        zscaler.fit(Xt)
        zscaler.transform_inplace(Xt)
        print_column_stats(label, Xt)
    print()

    # Step 3. LpNormScaler: L2 vs L1, with centering on and off (scale kept on).
    #         With scale=True each column is divided by its Lp norm around the
    #         applied center, so the average column L2 norm collapses toward 1
    #         for the L2 variant.
    l_cases = [
        (NormType.L2, True,  "LpNorm(L2, center=True )"),
        (NormType.L2, False, "LpNorm(L2, center=False)"),
        (NormType.L1, True,  "LpNorm(L1, center=True )"),
    ]
    for norm, center, label in l_cases:
        Xt = X.copy(order="F")
        lscaler = LpNormScaler(norm, center, True)
        lscaler.fit(Xt)
        lscaler.transform_inplace(Xt)
        print_column_stats(label, Xt)

    print("\n")


# ==============================================================================
# Demo 2: Transform and Inverse Transform Test
# ==============================================================================

def demo_scaler_inverse_transform():
    print_section_header("Scaler Inverse Transform Accuracy")

    # Step 1. Generate random data
    n, p = 1_000, 5_000
    print(f"Matrix size: {n} x {p}\n")

    rng = np.random.default_rng(456)
    X_orig = np.asfortranarray(rng.standard_normal((n, p)))

    # Step 2. Test ZScoreScaler
    X = X_orig.copy(order="F")
    zscaler = ZScoreScaler()
    zscaler.fit(X)
    zscaler.transform_inplace(X)
    zscaler.inverse_transform_inplace(X)

    err = np.abs(X - X_orig)
    print("ZScoreScaler:")
    print(f"  Max Error: {err.max():e}")
    print(f"  Mean Error: {err.mean():e}\n")

    # Step 3. Test LpNormScaler L2
    X = X_orig.copy(order="F")
    l2scaler = LpNormScaler(NormType.L2, True)
    l2scaler.fit(X)
    l2scaler.transform_inplace(X)
    l2scaler.inverse_transform_inplace(X)

    err = np.abs(X - X_orig)
    print("LpNormScaler L2:")
    print(f"  Max Error: {err.max():e}")
    print(f"  Mean Error: {err.mean():e}\n")
    print("\n")


# ==============================================================================
# Demo 3: Serialization
# ==============================================================================

def demo_scaler_serialization():
    print_section_header("Scaler Serialization")

    # Step 1. Generate data
    n, p = 1_000, 500
    rng = np.random.default_rng(999)
    X = np.asfortranarray(rng.standard_normal((n, p)))

    # Step 2. Fit and save (cpp: fallback filename "scaler_test.bin";
    #         a temporary file in the system temp directory is used here)
    fd, filename = tempfile.mkstemp(prefix="trex_scaler_demo_", suffix=".bin")
    os.close(fd)
    try:
        scaler = ZScoreScaler()
        scaler.fit(X)
        scaler.save(filename)
        print(f"✓ Saved ZScoreScaler to '{filename}'")

        # Step 3. Load and verify
        loaded_scaler = ZScoreScaler()
        loaded_scaler.load(filename)
        print(f"✓ Loaded ZScoreScaler from '{filename}'\n")

        # Verify means and scales match
        mean_diff = np.linalg.norm(scaler.get_centers() - loaded_scaler.get_centers())
        scale_diff = np.linalg.norm(scaler.get_scales() - loaded_scaler.get_scales())

        print("Verification:")
        print(f"  Mean difference: {mean_diff:e}")
        print(f"  Scale difference: {scale_diff:e}\n")

        if mean_diff < 1e-10 and scale_diff < 1e-10:
            print("✓ Serialization perfect!\n")
    finally:
        # Step 4. Clean up
        try:
            os.unlink(filename)
        except OSError:
            pass
    print("\n")


# ==============================================================================
# Demo 4: Memory-Mapped Matrix
# ==============================================================================

def demo_scaler_on_mmap_matrix():
    print_section_header("Scaler on Memory-Mapped Matrix")

    # Step 1. Create memory-mapped matrix
    #         (cpp: n = 20,000, p = 100,000; scaled to 2,000 x 10,000 here)
    print("Creating memory-mapped matrix...")
    n, p = 2_000, 10_000
    print(f"Matrix size: {n} x {p}")
    fd, mmap_filename = tempfile.mkstemp(prefix="trex_scaler_mmap_", suffix=".bin")
    os.close(fd)
    try:
        mmap_matrix = MemoryMappedMatrix(mmap_filename, n, p, AccessMode.ReadWrite)
        X_map = mmap_matrix.to_numpy()  # zero-copy view into the file

        # Step 2. Populate matrix: X(i, j) = 2*i + j/100
        #         (cpp fills the columns with an OpenMP parallel loop — that
        #         part is C++-exclusive; a vectorized broadcast is used here)
        i = np.arange(n, dtype=np.float64)[:, None]
        j = np.arange(p, dtype=np.float64)[None, :]
        X_map[:, :] = 2.0 * i + j / 100.0

        # Step 3. Fit and run LpNormScaler
        l2scaler = LpNormScaler(NormType.L2, True)
        print("Fitting L2 LpNormScaler on memory-mapped matrix...")
        l2scaler.fit(X_map)
        print("✓ LpNormScaler fitted")
        l2scaler.transform_inplace(X_map)
        print("✓ Data transformed")
        l2scaler.inverse_transform_inplace(X_map)
        print("✓ Data inverse transformed")
        del X_map, mmap_matrix  # unmap before the file is removed
    finally:
        # Step 4. Clean up
        try:
            os.unlink(mmap_filename)
        except OSError:
            pass
        print("✓ Memory-mapped file removed")
    print("\n")


# ==============================================================================
# Main
# ==============================================================================

if __name__ == "__main__":
    print()
    print_section_header("DataTransformer Demo Suite")

    # cpp initializes OpenMP with 6 threads; the backend thread pool is
    # configured here through the same utils entry points.
    set_num_threads(min(6, get_max_threads()))
    print(f"Running with {get_max_threads()} threads\n")

    demo_scaler_comparison()
    demo_scaler_inverse_transform()
    demo_scaler_serialization()
    demo_scaler_on_mmap_matrix()

    print()
    print_section_header("All Demos Completed Successfully")
